### PR TITLE
[luci] Allow non-constant weight for FC

### PR DESCRIPTION
**Pull request by AShedko**
This commit allows non-constant weight (B in AxB+C)
for FullyConnected Layer quantization.
This is needed in order to support Transformer-type models.

ONE-DCO-1.0-Signed-off-by: Andrey Shedko <a.shedko@samsung.com>

### CODE REVIEW

**Comment by jinevening** [comment]
I think this change looks dangerous. FC uses its own quantization scheme for weights, which is different from normal feature maps.

AFAIK, FC generated from transformer is originally matmul Op.  So, can you replace FC (w/ non-const weights) with BatchMatMul Op? Circle quantizer supports BatchMatMul quantization.

**Comment by AShedko** [comment]
> I think this change looks dangerous. FC uses its own quantization scheme for weights, which is different from normal feature maps.
> 
> AFAIK, FC generated from transformer is originally matmul Op. So, can you replace FC (w/ non-const weights) with BatchMatMul Op? Circle quantizer supports BatchMatMul quantization.

Do you suggest adding a special pass for this case?

**Comment by jinevening** [comment]
Yes. I think making a new pass is more appropriate in this case.  I'd like to avoid making an exception for quantizing weights. The reason is simple; FC is not for matmul.

I don't know the background of this pr, so please let me know if I'm missing something.

**Comment by AShedko** [comment]
@jinevening, PTAL?

File: compiler/luci/pass/include/luci/Pass/ReplaceNonConstFCWithBatchMatMulPass.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (AShedko) [inline compiler/luci/pass/include/luci/Pass/ReplaceNonConstFCWithBatchMatMulPass.h:2]:
```suggestion
 * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2020 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (AShedko) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:2]:
```suggestion
 * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (AShedko) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp:2]:
```suggestion
 * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ *       [transpose]?   [transpose]?
+ *               \        /
+ *           [ Fully Connected ]
+ *
+ *  AFTER
+ *
+ *              [Node1]  [Node2]
+ *                  \      /
+ *                [BatchMatMul]
+ *
```
Comment (AShedko) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:40]:
```suggestion
 *
 *              [Node1]  [Node2]
 *                  \      /
 *               [BatchMatMul]   [BiasValue?]
 *                          \       / 
 *                         [  Add?  ]
 *
 * Nodes with "?" denote optional elements
```

**Comment by seanshpark** [comment]
@AShedko , can you prepare a DRAFT PR?
- with this change and a sample network recipe and 
- a dredd test so that
- with this option and without this option is verified in execution result

**Comment by AShedko** [comment]
> @AShedko , can you prepare a DRAFT PR?
> 
> * with this change and a sample network recipe and
> * a dredd test so that
> * with this option and without this option is verified in execution result

FullyConnected_006 is suitable to show this change

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ *  Replace channel-wise Mul/Add with DepthwiseConv2D
+ *
+ *  BEFORE
+ *
+ *         [Node1]         [Node2]
+ *           |               |
+ *       [transpose]?   [transpose]?
+ *               \        /
+ *           [ Fully Connected ]
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:33]: minor thing, how about `[FullyConnected]` ?

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ *  AFTER
+ *
+ *              [Node1]  [Node2]
+ *                  \      /
+ *               [BatchMatMul]   [BiasValue?]
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:39]: minor thing, to match with above; `[BiasValue]?` ?

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+ *                          \       /
+ *                         [  Add?  ]
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:41]: `[ Add? ]` -> `[Add]?` ?

Comment (AShedko) [reply compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp]: The padding with spaces is for aesthetic reasons

Comment (seanshpark) [reply compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp]: Um? Can you please explain more? If you look other Pass comments `[]` and the `name` use same convention but I see different here.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ *  Replace channel-wise Mul/Add with DepthwiseConv2D
+ *
+ *  BEFORE
+ *
+ *         [Node1]         [Node2]
+ *           |               |
+ *       [transpose]?   [transpose]?
+ *               \        /
+ *           [ FullyConnected ]
```
Comment (AShedko) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:33]:
```suggestion
 *            [FullyConnected]
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ *  AFTER
+ *
+ *              [Node1]  [Node2]
+ *                  \      /
+ *               [BatchMatMul]   [BiasValue]?
+ *                        \         /
+ *                       [    Add    ]?
+ *
```
Comment (AShedko) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:42]:
```suggestion
 *               [BatchMatMul] [BiasValue]?
 *                        \       /
 *                          [Add]?
 *
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp (hunk 7 above)
Comment (AShedko) [reply compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp]: PTAL, padding removed

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+#include <luci/Pass/ReplaceNonConstFCWithBatchMatMulPass.h>
+
+namespace
+{
+
+/**
+ *  Replace channel-wise Mul/Add with DepthwiseConv2D
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:25]: This commnet seems need update

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    {
+      ac &= c->at<loco::DataType::FLOAT32>(i) == 0.0f;
+    }
+    return ac;
+  };
+
+  luci::CircleConst *bc;
+  if ((bc = dynamic_cast<luci::CircleConst *>(b)) && !all_zero(bc))
+  {
+    auto bias = fc->graph()->nodes()->create<luci::CircleAdd>();
```
Comment (seanshpark) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:111]: optional) although this Op is for bias, I think it would be better to rename the variable like `bias_add` or `add_bias` this op is `CircleAdd` for better readability.

**Comment by seanshpark** [comment]
> I don't know the background of this pr, so please let me know if I'm missing something.

I also am curious if FullyConnected can be replaced with BatchMatMul...
@AShedko , can you share any reference explaining this?

**Comment by AShedko** [comment]
> > I don't know the background of this pr, so please let me know if I'm missing something.
> 
> I also am curious if FullyConnected can be replaced with BatchMatMul... @AShedko , can you share any reference explaining this?

`FC(X,W,B) = MatMul(X,W.T)+B`

See https://keras.io/api/layers/core_layers/dense/ for example description. Also, converting ONNX models with MatMul op gives Fully Connected ops.

**Review by seanshpark — DISMISSED** [review]
LGTM thank you!

**Comment by AShedko** [comment]
@jinevening, PTAL

**Comment by AShedko** [comment]
@jinevening, PTAL?

**Comment by seanshpark** [comment]
@jinevening is on vacation. he sometimes visit here so I expected him to do so.
I think he'll be back at 21'st.
You can wait or I can land this without his approval but I hope to get review from him.

**Comment by AShedko** [comment]
> You can wait or I can land this without his approval but I hope to get review from him.

This is not urgent, I can wait for @jinevening.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    {
+      ac &= c->at<loco::DataType::FLOAT32>(i) == 0.0f;
+    }
+    return ac;
+  };
+
+  luci::CircleConst *bc;
+  if ((bc = dynamic_cast<luci::CircleConst *>(b)) && !all_zero(bc))
+  {
+    auto bias_add = fc->graph()->nodes()->create<luci::CircleAdd>();
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:113]: Origin of `bias_add` needs to be set.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+  auto all_zero = [](const luci::CircleConst *c) {
+    bool ac = true;
+    for (uint32_t i = 0; i < c->size<loco::DataType::FLOAT32>() && ac; i++)
+    {
+      ac &= c->at<loco::DataType::FLOAT32>(i) == 0.0f;
+    }
+    return ac;
+  };
+
+  luci::CircleConst *bc;
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:110]:
```suggestion
  luci::CircleConst *bc = nullptr;
```

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    bias_add->x(matmul);
+    bias_add->y(b);
+    bias_add->name(fc->name() + "/bias_add");
+    bias_add->dtype(fc->dtype());
+    loco::replace(fc).with(bias_add);
+    bias_add->fusedActivationFunction(fc->fusedActivationFunction());
+  }
+  else
+  {
+    loco::replace(fc).with(matmul);
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:123]: If fc has a fused activation function with no bias, the fused activation function will be lost.

We'll need to create an Op for fused activation function, or just ignore the case.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+{
+
+/**
+ *  Replace Fully Connected with Batched MatMul
+ *
+ *  BEFORE
+ *
+ *         [Node1]         [Node2]
+ *           |               |
+ *       [transpose]?   [transpose]?
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.cpp:31]: (optional) I think Transpose-BMM pattern can be optimized in another pass. Considering Transpose makes this pass complicated.

Please note that this is just an optional comment.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+ *
+ *  BEFORE
+ *
+ *         [FC]
+ *           |
+ *     [Add w/ Relu]
+ *
+ *  AFTER
+ *
+ *      [FC w/ Relu] (bias updated)
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp:86]: This comment does not match with the actual graph.

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+ *
+ */
+struct FCGraphlet
+{
+public:
+  FCGraphlet() = default;
+
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp:93]: Please add virtual destructor

File: compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+{
+  g.init({2, 3}, {2, 3}, {2, 2}, 1.0f);
+
+  auto ret = pass.run(g.g());
+  EXPECT_EQ(true, ret);
+
+  auto mm = dynamic_cast<luci::CircleAdd *>(g.output()->from());
+  EXPECT_NE(nullptr, mm);
+}
+
```
Comment (jinevening) [inline compiler/luci/pass/src/ReplaceNonConstFCWithBatchMatMulPass.test.cpp:167]: Could you add a value test?

**Comment by jinevening** [comment]
> This is not urgent, I can wait for @jinevening.

Sorry for the late review. I came back to work today.

**Review by jinevening — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM
